compiler/rustc_hir_analysis/src/collect/type_of.rs

```diff
@@ -693,6 +693,12 @@ fn find_opaque_ty_constraints_for_tait(tcx: TyCtxt<'_>, def_id: LocalDefId) -> T
         tcx.sess.emit_err(UnconstrainedOpaqueType {
             span: tcx.def_span(def_id),
             name: tcx.item_name(tcx.local_parent(def_id).to_def_id()),
+            what: match tcx.hir().get(scope) {
+                _ if scope == hir::CRATE_HIR_ID => "module",
+                Node::Item(hir::Item { kind: hir::ItemKind::Mod(_), .. }) => "module",
+                Node::Item(hir::Item { kind: hir::ItemKind::Impl(_), .. }) => "impl",
+                _ => "item",
```

Comment:
Not sure how to test this. Of would require mentioning the TAIT outside of a function body where it's defined. Maybe span_bug it?

Reply:
Are you talking about the `"item"` arm specifically? This is an example of where it would now print `"item"`:

```rust
#![feature(type_alias_impl_trait)]

fn foo() {
    type Tait = impl Sized;
}
```

I guess I could make it specifically match a few more cases where we could see TAITs like function bodies, and then `span_bug` the rest?

Reply:
Right... it maybe useful to special case all of these (e.g. the crate root mentioning the same crate (can we even get that error?))

But this is already correct now, so let's merge it